File: src/Extism.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

public abstract class ValueX
{
    [JsonIgnore]
    public abstract string ValueType { get; }
}

public class ValueDateTime : ValueX
{
    [JsonPropertyName("valueDateTime")]
    public required string Value { get; set; }

    [JsonIgnore]
    public override string ValueType => "valueDateTime";
}


public class ValueString : ValueX
{
    [JsonPropertyName("valueString")]
    public required string Value { get; set; }

    [JsonIgnore]
    public override string ValueType => "valueString";
}

public class ValueCode : ValueX
{
    [JsonPropertyName("valueCode")]
    public required string Value { get; set; }

    [JsonIgnore]
    public override string ValueType => "valueCode";
}

public class ValueCoding : ValueX
{
    [JsonPropertyName("valueCoding")]
    public required Coding Value { get; set; }

    [JsonIgnore]
    public override string ValueType => "valueCoding";
}

public class ValueDecimal : ValueX
{
    [JsonPropertyName("valueDecimal")]
    public required string Value { get; set; }

    [JsonIgnore]
    public override string ValueType => "valueDecimal";
}

public class Coding
{
    [JsonPropertyName("system")]
    public string? System { get; set; }

    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonPropertyName("display")]
    public string? Display { get; set; }
}

[JsonConverter(typeof(PropertyConverter))]
public class Property
{
    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonIgnore]
    public ValueX? Value { get; set; }
}

public class PropertyConverter : JsonConverter<Property>
{
    public override Property Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        var property = new Property();
        while (reader.Read())
        {
            if (reader.TokenType == JsonTokenType.EndObject)
                break;

            if (reader.TokenType == JsonTokenType.PropertyName)
            {
                var propertyName = reader.GetString();
                reader.Read();

                switch (propertyName)
                {
                    case "code":
                        property.Code = reader.GetString();
                        break;
                    case "valueString":
                        property.Value = new ValueString { Value = reader.GetString() ?? "" };
                        break;
                    case "valueDateTime":
                        property.Value = new ValueDateTime { Value = reader.GetString() ?? "" };
                        break;
                    case "valueCode":
                        property.Value = new ValueCode { Value = reader.GetString() ?? "" };
                        break;
                    case "valueCoding":
                        property.Value = new ValueCoding { Value = JsonSerializer.Deserialize<Coding>(ref reader, options) ?? new Coding() };
                        break;
                    case "valueDecimal":
                        property.Value = new ValueDecimal { Value = reader.GetString() ?? "" };
                        break;
                }
            }
        }
        return property;
    }

    public override void Write(Utf8JsonWriter writer, Property value, JsonSerializerOptions options)
    {
        writer.WriteStartObject();
        writer.WriteString("code", value.Code);

        switch (value?.Value?.ValueType)
        {
            case "valueString":
                writer.WriteString("valueString", ((ValueString)value.Value).Value);
                break;
            case "valueDateTime":
                writer.WriteString("valueDateTime", ((ValueDateTime)value.Value).Value);
                break;
            case "valueCode":
                writer.WriteString("valueCode", ((ValueCode)value.Value).Value);
                break;
            case "valueCoding":
                writer.WritePropertyName("valueCoding");
                JsonSerializer.Serialize(writer, ((ValueCoding)value.Value).Value, options);
                break;
            case "valueDecimal":
                writer.WriteString("valueDecimal", ((ValueDecimal)value.Value).Value);
                break;
        }

        writer.WriteEndObject();
    }
}

public class Concept
{
    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonPropertyName("display")]
    public string? Display { get; set; }


    [JsonPropertyName("properties")]
    public List<Property> Properties { get; set; } = new List<Property>();
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum Severity
{
    [JsonPropertyName("error")]
    Error,

    [JsonPropertyName("warning")]
    Warning,

    [JsonPropertyName("information")]
    Information,

    [JsonPropertyName("success")]
    Success
}

[JsonConverter(typeof(ParseDetailConverter))]
public class ParseDetail
{
    [JsonPropertyName("severity")]
    public Severity Severity { get; set; }

    [JsonPropertyName("key")]
    public string? Key { get; set; }

    [JsonIgnore]
    public ValueX? Value { get; set; }
}

public class ParseDetailConverter : JsonConverter<ParseDetail>
{
    public override ParseDetail Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        var parseDetail = new ParseDetail();
        while (reader.Read())
        {
            if (reader.TokenType == JsonTokenType.EndObject)
                break;

            if (reader.TokenType == JsonTokenType.PropertyName)
            {
                var propertyName = reader.GetString();
                reader.Read();

                switch (propertyName)
                {
                    case "severity":
                        parseDetail.Severity = JsonSerializer.Deserialize<Severity>(ref reader, options);
                        break;
                    case "key":
                        parseDetail.Key = reader.GetString();
                        break;
                    case "valueString":
                        parseDetail.Value = new ValueString { Value = reader.GetString() ?? "" };
                        break;
                    case "valueCode":
                        parseDetail.Value = new ValueCode { Value = reader.GetString() ?? "" };
                        break;
                    case "valueCoding":
                        parseDetail.Value = new ValueCoding { Value = JsonSerializer.Deserialize<Coding>(ref reader, options) ?? new Coding() };
                        break;
                    case "valueDecimal":
                        parseDetail.Value = new ValueDecimal { Value = reader.GetString() ?? "" };
                        break;
                }
            }
        }
        return parseDetail;
    }

    public override void Write(Utf8JsonWriter writer, ParseDetail value, JsonSerializerOptions options)
    {
        writer.WriteStartObject();
        writer.WriteString("severity", value.Severity.ToString().ToLowerInvariant());
        writer.WriteString("key", value.Key);

        switch (value?.Value?.ValueType)
        {
            case "valueString":
                writer.WriteString("valueString", ((ValueString)value.Value).Value);
                break;
            case "valueCode":
                writer.WriteString("valueCode", ((ValueCode)value.Value).Value);
                break;
            case "valueCoding":
                writer.WritePropertyName("valueCoding");
                JsonSerializer.Serialize(writer, ((ValueCoding)value.Value).Value, options);
                break;
            case "valueDecimal":
                writer.WriteString("valueDecimal", ((ValueDecimal)value.Value).Value);
                break;
        }

        writer.WriteEndObject();
    }
}

public class ParseResponse
{
    [JsonPropertyName("details")]
    public List<ParseDetail>? Details { get; set; }

    [JsonPropertyName("concept")]
    public Concept? Concept { get; set; }
}

// TerminologyDb.cs

public class LookupRequest
{
    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonPropertyName("properties")]
    public List<string>? Properties { get; set; }
}

public class LookupResponse
{
    [JsonPropertyName("concept")]
    public Concept? Concept { get; set; }
}

public class SubsumesRequest
{
    [JsonPropertyName("ancestor")]
    public string? Ancestor { get; set; }

    [JsonPropertyName("descendant")]
    public string? Descendant { get; set; }
}

public class SubsumesResponse
{
    [JsonPropertyName("subsumes")]
    public bool Subsumes { get; set; }
}

// TerminologyEngine.cs

public class ParseRequest
{
    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonPropertyName("properties")]
    public List<string>? Properties { get; set; }
}

File: bcp47/src/extism.rs
use extism_pdk::{host_fn, info, log, plugin_fn, FnResult, LogLevel};
use extism_pdk::{FromBytes, Json, ToBytes};
use nom::error::ParseError;
use serde_derive::{Deserialize, Serialize};

pub struct Guest<T>
where
    T: TerminologyDb,
{
    pub(crate) db: T,
}

impl Default for Guest<HostReal> {
    fn default() -> Self {
        Guest::new(HostReal::new())
    }
}

impl<T> Guest<T>
where
    T: TerminologyDb,
{
    pub(crate) fn new(host: T) -> Self {
        Guest { db: host }
    }
}

impl ParseError<&str> for ParseDetail {
    fn from_error_kind(input: &str, kind: nom::error::ErrorKind) -> Self {
        ParseDetail {
            severity: Severity::Error,
            key: "error".to_string(),
            value: ValueX::ValueString(format!(
                "Error parsing input: {:?} with kind: {:?}",
                input, kind
            )),
        }
    }

    fn append(input: &str, kind: nom::error::ErrorKind, _other: Self) -> Self {
        ParseDetail {
            severity: Severity::Error,
            key: "error".to_string(),
            value: ValueX::ValueString(format!(
                "Error parsing input: {:?} with kind: {:?}",
                input, kind
            )),
        }
    }
}

pub trait TerminologyEngine<T: TerminologyDb> {
    fn metadata(&self) -> String;
    fn parse(&self, req: ParseRequest) -> ParseResponse;
    fn subsumes(&self, req: SubsumesRequest) -> SubsumesResponse;
}

pub trait TerminologyDb {
    fn db_lookup(&self, req: LookupRequest) -> LookupResponse;
    fn db_subsumes(&self, req: SubsumesRequest) -> SubsumesResponse;
}

pub trait WithDb {
    fn db(&self) -> &dyn TerminologyDb;
}

#[derive(Copy, Clone)]
pub struct HostReal {}

impl HostReal {
    pub fn new() -> Self {
        HostReal {}
    }
}

#[host_fn]
extern "ExtismHost" {
    fn db_lookup(input: LookupRequest) -> LookupResponse;
}

#[host_fn]
extern "ExtismHost" {
    fn db_subsumes(input: LookupRequest) -> LookupResponse;
}
impl TerminologyDb for HostReal {
    fn db_lookup(&self, req: LookupRequest) -> LookupResponse {
        log!(LogLevel::Info, "Calling host");
        unsafe {
            let res = db_lookup(req);
            info!("Got back : {:?}", res);
            res.unwrap_or(LookupResponse { concept: None })
        }
    }

    fn db_subsumes(&self, _req: SubsumesRequest) -> SubsumesResponse {
        todo!()
    }
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub enum ValueX {
    #[serde(rename = "valueString")]
    ValueString(String),
    #[serde(rename = "valueDateTime")]
    ValueDateTime(String),
    #[serde(rename = "valueCode")]
    ValueCode(String),
    #[serde(rename = "valueCoding")]
    ValueCoding(Coding),
    #[serde(rename = "valueDecimal")]
    ValueDecimal(String),
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct Coding {
    pub system: Option<String>,
    pub code: Option<String>,
    pub display: Option<String>,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct Property {
    pub code: String,
    #[serde(flatten)]
    pub value: ValueX,
}
#[derive(
    Default, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct Concept {
    pub code: String,
    pub display: Option<String>,
    pub properties: Vec<Property>,
}


#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Error,
    Warning,
    Information,
    Success,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct ParseDetail {
    pub severity: Severity,
    pub key: String,
    #[serde(flatten)]
    pub value: ValueX,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct ParseResponse {
    pub details: Vec<ParseDetail>,
    pub concept: Option<Concept>,
}

// terminology_db.rs

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct LookupRequest {
    pub code: String,
    pub properties: Option<Vec<String>>,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct LookupResponse {
    pub concept: Option<Concept>,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct SubsumesRequest {
    pub ancestor: String,
    pub descendant: String,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct SubsumesResponse {
    pub subsumes: bool,
}

// terminology_engine.rs

#[derive(
    Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord, Clone, Debug, ToBytes, FromBytes,
)]
#[encoding(Json)]
pub struct ParseRequest {
    pub code: String,
    pub properties: Option<Vec<String>>,
}

pub use lazy_static::lazy_static;

#[macro_export]
macro_rules! define_terminology_engine {
    ($engine_type:ty) => {
        lazy_static! {
            static ref TERMINOLOGY_ENGINE: $engine_type = <$engine_type>::default();
        }

        #[plugin_fn]
        pub fn metadata() -> FnResult<String> {
            Ok(TERMINOLOGY_ENGINE.metadata())
        }

        #[plugin_fn]
        pub fn parse(req: ParseRequest) -> FnResult<ParseResponse> {
            Ok(TERMINOLOGY_ENGINE.parse(req))
        }

        #[plugin_fn]
        pub fn subsumes(req: SubsumesRequest) -> FnResult<SubsumesResponse> {
            Ok(TERMINOLOGY_ENGINE.subsumes(req))
        }
    };
}

define_terminology_engine!(Guest<HostReal>);

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::to_string;

    #[test]
    fn test_parse_request_response() {
        let request = ParseRequest {
            code: "test_code".to_string(),
            properties: Some(vec!["property1".to_string(), "property2".to_string()]),
            // Add more properties here as needed
        };

        let response = ParseResponse {
            details: vec![
                ParseDetail {
                    severity: Severity::Error,
                    key: "key1".to_string(),
                    value: ValueX::ValueCode("OK".to_string()),
                },
                // Add more Detail structs here as needed
            ],
            concept: Some(Concept {
                code: "concept_code".to_string(),
                properties: vec![
                    Property {
                        code: "property1".to_string(),
                        value: ValueX::ValueString("value1".to_string()),
                    },
                    Property {
                        code: "property2".to_string(),
                        value: ValueX::ValueCoding(Coding {
                            system: Some("system1".to_string()),
                            code: Some("code1".to_string()),
                            display: Some("display1".to_string()),
                            // Add more properties here as needed
                        }),
                    },
                    // Add more Property structs here as needed
                ],
                ..Concept::default()
            }),
        };

        let request_json = to_string(&request).unwrap();
        let response_json = to_string(&response).unwrap();

        println!("Request: {}", request_json);
        println!("Response: {}", response_json);
    }
}
